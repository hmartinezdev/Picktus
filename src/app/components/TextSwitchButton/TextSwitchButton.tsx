import colors from '@constants/colors';
import React, { Component } from 'react';
import { Transition } from 'react-transition-group';

export interface ITextSwitchButtonPropTypes {
  activeText: string;
  initialText: string;
  onClick(): void;
}

export interface ITextSwitchButtonState {
  active: boolean;
}

class Button extends Component<ITextSwitchButtonPropTypes, ITextSwitchButtonState> {
  public static defaultProps = {
    activeText: '',
    initialText: '',
  };

  public constructor(props: ITextSwitchButtonPropTypes) {
    super(props);

    this.state = {
      active: false,
    };
  }

  public onClick = (): void => {
    this.setState({ active: !this.state.active });
    this.props.onClick();
  };

  public render(): React.ReactElement<Button> {
    const { initialText, activeText } = this.props;
    const { active } = this.state;

    return (
      <button onClick={this.onClick} className={`button ${active ? 'button--active' : ''}`}>
        <Transition key="initial" timeout={200} in={!active} mountOnEnter={true} unmountOnExit={true}>
          {(status) => <p className={`text text--${status}`}>{initialText}</p>}
        </Transition>
        <Transition key="active" timeout={200} in={active} mountOnEnter={true} unmountOnExit={true}>
          {(status) => <p className={`text text--${status}`}>{activeText}</p>}
        </Transition>
        <style jsx>{`
          .button {
            font-size: 0.9rem;
            line-height: 2.5rem;
            background-color: ${colors.secondary}89;
            padding: 0 0.8rem;
            height: 2.3rem;
            font-family: 'Josefin Sans', sans-serif;
            color: ${colors.primary};
            border-radius: 3px;
            margin-bottom: 0.8rem;
            transition: all ease-in 200ms;
            border: none;
            width: 100%;
            outline: none;
            cursor: pointer;
            position: relative;
            overflow: hidden;
          }

          .button:hover {
            background-color: ${colors.secondary};
          }

          .button--active {
            background-color: ${colors.primary}89;
            color: ${colors.secondary};
          }

          .button--active:hover {
            background-color: ${colors.primary};
          }

          .text {
            position: absolute;
            margin: 0;
            padding: 0;
            left: 50%;
            top: 50%;
            transform: translate(-50%, 50%);
            width: 100%;
            cursor: pointer;
          }

          .text--exiting,
          .text--exited {
            transition: all 200ms ease-in;
            transform: translate(-50%, 100%);
          }

          .text--entering {
            animation: slidedown 200ms ease-in;
          }

          .text--entered {
            transform: translate(-50%, -50%);
          }

          @keyframes slidedown {
            0% {
              transform: translate(-50%, -150%);
            }

            100% {
              transform: translate(-50%, -50%);
            }
          }
        `}</style>
      </button>
    );
  }
}

export default Button;
