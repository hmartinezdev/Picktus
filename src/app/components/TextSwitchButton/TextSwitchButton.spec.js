import { shallow } from 'enzyme';
import React from 'react';
import TextSwitchButton from './TextSwitchButton';

let wrapper;
const basicProps = {
  onClick: () => {},
  activeText: 'testactive',
  initialText: 'testinitial',
};

const setup = (props = basicProps) => shallow(<TextSwitchButton {...basicProps} {...props} />);

describe('<TextSwitchButton />', () => {
  beforeEach(() => {
    wrapper = setup();
  });

  test('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('it should render the second text properly', () => {
    wrapper.setState({ active: true });
    expect(wrapper).toMatchSnapshot();
  });

  describe('onClick', () => {
    test('it should toggle active value', () => {
      const mock = jest.fn();
      wrapper = setup({ onClick: mock });
      const instance = wrapper.instance();
      const spy = jest.spyOn(instance, 'setState');
      instance.onClick();
      expect(spy).toHaveBeenCalledWith({ active: true });
      expect(mock).toHaveBeenCalled();
    });
  });
});
