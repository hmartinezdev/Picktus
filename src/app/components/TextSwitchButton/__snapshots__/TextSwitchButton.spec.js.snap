// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextSwitchButton /> should render properly 1`] = `
<button
  className="button"
  onClick={[Function]}
>
  <Transition
    appear={false}
    enter={true}
    exit={true}
    in={true}
    key="initial"
    mountOnEnter={true}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={200}
    unmountOnExit={true}
  >
    <Component />
  </Transition>
  <Transition
    appear={false}
    enter={true}
    exit={true}
    in={false}
    key="active"
    mountOnEnter={true}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    timeout={200}
    unmountOnExit={true}
  >
    <Component />
  </Transition>
  <style
    jsx={true}
  >
    
          .button {
            font-size: 0.9rem;
            line-height: 2.5rem;
            background-color: #FAC9B889;
            padding: 0 0.8rem;
            height: 2.3rem;
            font-family: 'Josefin Sans', sans-serif;
            color: #2d3047;
            border-radius: 3px;
            margin-bottom: 0.8rem;
            transition: background-color ease-in 200ms;
            border: none;
            width: 100%;
            outline: none;
            cursor: pointer;
            position: relative;
            overflow: hidden;
          }

          .button:hover {
            background-color: #FAC9B8cd;
          }

          .text {
            position: absolute;
            margin: 0;
            padding: 0;
            left: 50%;
            top: 50%;
            transform: translate(-50%, 50%);
            width: 100%;
            cursor: pointer;
          }

          .text--exiting,
          .text--exited {
            transition: all 200ms ease-in;
            transform: translate(-50%, 100%);
          }

          .text--entering {
            animation: slidedown 200ms ease-in;
          }

          .text--entered {
            transform: translate(-50%, -50%);
          }

          @keyframes slidedown {
            0% {
              transform: translate(-50%, -150%);
            }

            100% {
              transform: translate(-50%, -50%);
            }
          }
        
  </style>
</button>
`;
